File: src/backend/OutlierCredentialsManager.ts
import fs from "fs";
import path from "path";
import crypto from "crypto";
import dotenv from "dotenv";

export interface LoginResult {
  statusCode: number;
  body: unknown;
  cookie: string;
  csrf: string;
}

// ── Encryption (AES-256-GCM, key from SSH private key) ──────────────────────

const ALGO = "aes-256-gcm";

function deriveKey(sshKeyPath: string, salt: string): Buffer {
  if (!fs.existsSync(sshKeyPath)) throw new Error(`SSH key not found: ${sshKeyPath}`);
  return crypto.pbkdf2Sync(fs.readFileSync(sshKeyPath), salt, 100_000, 32, "sha512");
}

function encrypt(plaintext: string, key: Buffer): string {
  const iv = crypto.randomBytes(16);
  const cipher = crypto.createCipheriv(ALGO, key, iv);
  const data = Buffer.concat([cipher.update(plaintext, "utf8"), cipher.final()]);
  return [iv, cipher.getAuthTag(), data].map(b => b.toString("base64")).join(":");
}

function decrypt(packed: string, key: Buffer): string {
  const [iv, tag, data] = packed.split(":").map(s => Buffer.from(s, "base64"));
  if (!iv || !tag || !data) throw new Error("Malformed encrypted value");
  const d = crypto.createDecipheriv(ALGO, key, iv);
  d.setAuthTag(tag);
  return Buffer.concat([d.update(data), d.final()]).toString("utf8");
}

// ── Env file helpers ─────────────────────────────────────────────────────────

function readEnv(filepath: string): Record<string, string> {
  return fs.existsSync(filepath) ? dotenv.parse(fs.readFileSync(filepath)) : {};
}

function writeEnv(filepath: string, data: Record<string, string>) {
  const lines = Object.entries(data)
      .map(([k, v]) => `${k}="${v.replace(/"/g, '\\"')}"`)
      .join("\n");
  fs.writeFileSync(filepath, lines + "\n");
}

// ── Manager ──────────────────────────────────────────────────────────────────

export class OutlierCredentialsManager {
  static readonly OUTLIER_BASE_URL = "https://app.outlier.ai";
  static readonly REMOTASKS_BASE_URL = "https://www.remotasks.com";

  private readonly envPath: string;
  private readonly store: string;
  private suffix: string;
  private cookie: string | null = null;
  private csrf: string | null = null;

  constructor(store = "default_user", folder = "./env", suffix = "") {
    this.store = store;
    this.suffix = suffix;
    this.envPath = path.join(folder, `${store}.env`);
    if (!fs.existsSync(folder)) fs.mkdirSync(folder, { recursive: true });
    this.loadSession();
  }

  // key("COOKIE") → "COOKIE" or "COOKIE_prod"
  private key(name: string) {
    return this.suffix ? `${name}_${this.suffix}` : name;
  }

  private get salt() {
    return `outlier-creds:${this.store}:${this.suffix}`;
  }

  private sshKey(override?: string) {
    if (override) return deriveKey(override, this.salt);
    const sshDir = path.join(process.env.HOME ?? process.env.USERPROFILE ?? "~", ".ssh");
    for (const name of ["id_rsa", "id_ed25519", "id_ecdsa", "id_ecdsa_sk", "id_ed25519_sk"]) {
      const p = path.join(sshDir, name);
      if (fs.existsSync(p)) return deriveKey(p, this.salt);
    }
    throw new Error(`No SSH private key found in ${sshDir}`);
  }

  // ── Session ────────────────────────────────────────────────────────────

  private loadSession() {
    const cfg = readEnv(this.envPath);
    this.cookie = cfg[this.key("COOKIE")] ?? null;
    this.csrf = cfg[this.key("CSRF")] ?? null;
  }

  private saveSession(_username: string, cookie: string, csrf: string) {
    const cfg = readEnv(this.envPath);
    cfg[this.key("COOKIE")] = cookie;
    cfg[this.key("CSRF")] = csrf;
    writeEnv(this.envPath, cfg);
  }

  getCookie() { return this.cookie; }
  getCsrf() { return this.csrf; }
  getEnvPath() { return this.envPath; }
  hasCredentials() { return !!(this.cookie && this.csrf); }

  setSuffix(s: string) {
    this.suffix = s;
    this.loadSession();
  }

  // ── Credential storage ─────────────────────────────────────────────────

  storeCredentials(email: string, password: string, sshKeyPath?: string) {
    const k = this.sshKey(sshKeyPath);
    const cfg = readEnv(this.envPath);
    cfg[this.key("ENC_EMAIL")] = encrypt(email, k);
    cfg[this.key("ENC_PASSWORD")] = encrypt(password, k);
    writeEnv(this.envPath, cfg);
  }

  getStoredCredentials(sshKeyPath?: string) {
    const cfg = readEnv(this.envPath);
    const enc = cfg[this.key("ENC_EMAIL")];
    const pwd = cfg[this.key("ENC_PASSWORD")];
    if (!enc || !pwd) return null;

    try {
      const k = this.sshKey(sshKeyPath);
      return { email: decrypt(enc, k), password: decrypt(pwd, k) };
    } catch (err) {
      console.error("[credentials] decryption failed:", err);
      return null;
    }
  }

  clear() {
    const cfg = readEnv(this.envPath);
    for (const name of ["COOKIE", "CSRF", "ENC_EMAIL", "ENC_PASSWORD"]) {
      delete cfg[this.key(name)];
    }
    writeEnv(this.envPath, cfg);
    this.cookie = null;
    this.csrf = null;
  }

  // ── HTTP ───────────────────────────────────────────────────────────────

  private async request(
      method: string, endpoint: string,
      data?: Record<string, unknown>, useOutlier = true, attempt = 1,
  ): Promise<Response | null> {
    const base = useOutlier
        ? OutlierCredentialsManager.OUTLIER_BASE_URL
        : OutlierCredentialsManager.REMOTASKS_BASE_URL;
    const url = `${base}${endpoint}`;

    const headers: Record<string, string> = {
      "accept": "*/*",
      "accept-language": "en-US,en-GB;q=0.9,en;q=0.8",
      "cache-control": "no-cache",
      "content-type": "application/json",
      "user-agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36",
    };

    try {
      const opts: RequestInit = { method, headers };
      if (data) opts.body = JSON.stringify(data);
      const resp = await fetch(url, opts);
      if (resp.status >= 500) console.error(`[request] ${resp.status} ${resp.statusText}`);
      return resp;
    } catch (err) {
      console.error(`[request] network error:`, err);
      if (attempt < 2) {
        await new Promise(r => setTimeout(r, Math.min(2 ** attempt * 1000, 10_000)));
        return this.request(method, endpoint, data, useOutlier, attempt + 1);
      }
      return null;
    }
  }

  // ── Login ──────────────────────────────────────────────────────────────

  private parseCookies(resp: Response): Record<string, string> {
    const jar: Record<string, string> = {};
    const raw: string[] = resp.headers.getSetCookie?.() ?? [];
    if (!raw.length) {
      const h = resp.headers.get("set-cookie");
      if (h) raw.push(...h.split(/,(?=\s*\w+=)/));
    }
    for (const entry of raw) {
      const pair = entry.split(";")[0] ?? "";
      const eq = pair.indexOf("=");
      if (eq > 0) jar[pair.substring(0, eq).trim()] = pair.substring(eq + 1).trim();
    }
    return jar;
  }

  private async doLogin(email: string, password: string, useOutlier: boolean): Promise<LoginResult | null> {
    const endpoint = useOutlier ? "/internal/loginNext/expert" : "/internal/loginNext/worker";
    const resp = await this.request("POST", endpoint, { email, password }, useOutlier);
    if (!resp) return null;

    let body: unknown = null;
    try { body = await resp.json(); } catch { /* empty */ }

    const jar = this.parseCookies(resp);
    const cookie = Object.entries(jar).map(([k, v]) => `${k}=${v}`).join("; ");
    const csrf = decodeURIComponent(jar["_csrf"] ?? "");

    if (cookie) {
      this.saveSession(email, cookie, csrf);
      this.loadSession();
    }

    return { statusCode: resp.status, body, cookie, csrf };
  }

  // two clear entry points, no overload gymnastics
  async login(email: string, password: string, useOutlier = true) {
    return this.doLogin(email, password, useOutlier);
  }

  async loginWithStoredCredentials(useOutlier = true, sshKeyPath?: string) {
    const creds = this.getStoredCredentials(sshKeyPath);
    if (!creds) {
      console.error(`[login] no stored credentials (suffix="${this.suffix}"). Call storeCredentials() first.`);
      return null;
    }
    return this.doLogin(creds.email, creds.password, useOutlier);
  }
}

File: src/backend/outlier_credentials_manager.py
import asyncio
import base64
import logging
import os
from pathlib import Path
from urllib.parse import unquote

import httpx
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
from dotenv import dotenv_values

logger = logging.getLogger(__name__)

OUTLIER_BASE_URL = "https://app.outlier.ai"
REMOTASKS_BASE_URL = "https://www.remotasks.com"

_UA = (
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
    "(KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36"
)


def _derive_key(ssh_key_path: str, salt: str) -> bytes:
    kdf = PBKDF2HMAC(algorithm=hashes.SHA512(), length=32, salt=salt.encode(), iterations=100_000)
    return kdf.derive(Path(ssh_key_path).read_bytes())


def _encrypt(plaintext: str, key: bytes) -> str:
    iv = os.urandom(16)
    encryptor = Cipher(algorithms.AES(key), modes.GCM(iv)).encryptor()
    ct = encryptor.update(plaintext.encode()) + encryptor.finalize()
    tag = encryptor.tag
    return ":".join(base64.b64encode(x).decode() for x in [iv, tag, ct])


def _decrypt(packed: str, key: bytes) -> str:
    iv_b, tag_b, ct_b = [base64.b64decode(p) for p in packed.split(":")]
    decryptor = Cipher(algorithms.AES(key), modes.GCM(iv_b, tag_b)).decryptor()
    return (decryptor.update(ct_b) + decryptor.finalize()).decode()


def _read_env(filepath: str) -> dict[str, str]:
    if not Path(filepath).exists():
        return {}
    return {k: v for k, v in dotenv_values(filepath).items() if v is not None}


def _write_env(filepath: str, data: dict[str, str]) -> None:
    content = "\n".join(
        '{}="{}"'.format(k, v.replace('"', '\\"')) for k, v in data.items()
    ) + "\n"
    Path(filepath).write_text(content)


class OutlierCredentialsManager:
    def __init__(self, store: str = "default_user", folder: str = "./env", suffix: str = ""):
        self.store = store
        self.suffix = suffix
        self.env_path = str(Path(folder) / f"{store}.env")
        Path(folder).mkdir(parents=True, exist_ok=True)
        self._cookie: str | None = None
        self._csrf: str | None = None
        self._load_session()

    def _key(self, name: str) -> str:
        return f"{name}_{self.suffix}" if self.suffix else name

    @property
    def _salt(self) -> str:
        return f"outlier-creds:{self.store}:{self.suffix}"

    def _ssh_key(self, override: str | None = None) -> bytes:
        if override:
            return _derive_key(override, self._salt)
        ssh_dir = (
            Path(os.environ.get("HOME") or os.environ.get("USERPROFILE") or "~").expanduser() / ".ssh"
        )
        for name in ["id_rsa", "id_ed25519", "id_ecdsa", "id_ecdsa_sk", "id_ed25519_sk"]:
            p = ssh_dir / name
            if p.exists():
                return _derive_key(str(p), self._salt)
        raise FileNotFoundError(f"No SSH private key found in {ssh_dir}")

    def _load_session(self) -> None:
        cfg = _read_env(self.env_path)
        self._cookie = cfg.get(self._key("COOKIE"))
        self._csrf = cfg.get(self._key("CSRF"))

    def _save_session(self, cookie: str, csrf: str) -> None:
        cfg = _read_env(self.env_path)
        cfg[self._key("COOKIE")] = cookie
        cfg[self._key("CSRF")] = csrf
        _write_env(self.env_path, cfg)

    def get_cookie(self) -> str | None:
        return self._cookie

    def get_csrf(self) -> str | None:
        return self._csrf

    def has_credentials(self) -> bool:
        return bool(self._cookie and self._csrf)

    def store_credentials(self, email: str, password: str, ssh_key_path: str | None = None) -> None:
        k = self._ssh_key(ssh_key_path)
        cfg = _read_env(self.env_path)
        cfg[self._key("ENC_EMAIL")] = _encrypt(email, k)
        cfg[self._key("ENC_PASSWORD")] = _encrypt(password, k)
        _write_env(self.env_path, cfg)

    def get_stored_credentials(self, ssh_key_path: str | None = None) -> dict[str, str] | None:
        cfg = _read_env(self.env_path)
        enc_email = cfg.get(self._key("ENC_EMAIL"))
        enc_pwd = cfg.get(self._key("ENC_PASSWORD"))
        if not enc_email or not enc_pwd:
            return None
        try:
            k = self._ssh_key(ssh_key_path)
            return {"email": _decrypt(enc_email, k), "password": _decrypt(enc_pwd, k)}
        except Exception as e:
            logger.error("Decryption failed: %s", e)
            return None

    def clear(self) -> None:
        cfg = _read_env(self.env_path)
        for name in ["COOKIE", "CSRF", "ENC_EMAIL", "ENC_PASSWORD"]:
            cfg.pop(self._key(name), None)
        _write_env(self.env_path, cfg)
        self._cookie = None
        self._csrf = None

    def _parse_cookies(self, headers: httpx.Headers) -> dict[str, str]:
        jar: dict[str, str] = {}
        for value in headers.get_list("set-cookie"):
            pair = value.split(";")[0]
            eq = pair.find("=")
            if eq > 0:
                jar[pair[:eq].strip()] = pair[eq + 1:].strip()
        return jar

    async def _request(
        self,
        method: str,
        endpoint: str,
        data: dict | None = None,
        use_outlier: bool = True,
        attempt: int = 1,
    ) -> httpx.Response | None:
        base = OUTLIER_BASE_URL if use_outlier else REMOTASKS_BASE_URL
        headers = {
            "accept": "*/*",
            "accept-language": "en-US,en-GB;q=0.9,en;q=0.8",
            "cache-control": "no-cache",
            "content-type": "application/json",
            "user-agent": _UA,
        }
        try:
            async with httpx.AsyncClient() as client:
                resp = await client.request(method, f"{base}{endpoint}", headers=headers, json=data)
                if resp.status_code >= 500:
                    logger.error("[request] %s %s", resp.status_code, resp.reason_phrase)
                return resp
        except httpx.RequestError as e:
            logger.error("[request] network error: %s", e)
            if attempt < 2:
                await asyncio.sleep(min(2**attempt, 10))
                return await self._request(method, endpoint, data, use_outlier, attempt + 1)
            return None

    async def _do_login(self, email: str, password: str, use_outlier: bool) -> dict | None:
        endpoint = "/internal/loginNext/expert" if use_outlier else "/internal/loginNext/worker"
        resp = await self._request("POST", endpoint, {"email": email, "password": password}, use_outlier)
        if not resp:
            return None

        try:
            body = resp.json()
        except Exception:
            body = None

        jar = self._parse_cookies(resp.headers)
        cookie = "; ".join(f"{k}={v}" for k, v in jar.items())
        csrf = unquote(jar.get("_csrf", ""))

        if cookie:
            self._save_session(cookie, csrf)
            self._load_session()

        return {"statusCode": resp.status_code, "body": body, "cookie": cookie, "csrf": csrf}

    async def login(self, email: str, password: str, use_outlier: bool = True) -> dict | None:
        return await self._do_login(email, password, use_outlier)

    async def login_with_stored_credentials(
        self, use_outlier: bool = True, ssh_key_path: str | None = None
    ) -> dict | None:
        creds = self.get_stored_credentials(ssh_key_path)
        if not creds:
            logger.error('[login] no stored credentials (suffix="%s")', self.suffix)
            return None
        return await self._do_login(creds["email"], creds["password"], use_outlier)
